src/math/BoundedField.cc

```diff
@@ -83,19 +83,110 @@ struct Divide {
+        // iterate over (y0, y1) intervals, then (x0, x1) intervals to evaluate
+        // the field at the corners of each box, then iterate over pixels in
+        // the box (in y, then x) to fill interpolated values.
+        // This means we evaluate the function multiple times at some points
+        // (when we move to a new interval in y, to be specific).
+        // Could probably be optimized for the cases where either step is 1,
+        // but that seems premature.
+
+        // Local function to do bilinear interpolation using points
+        // [(x0, y0), (x0, y1), (x1, y0), (x1, y1)] with values
+        // [z00, z01, z10, z11], in pixels [x0:xEnd, y0:yEnd]
+        auto interpolate = [&img, &functor](
+            int x0, int x1, int xEnd,
+            int y0, int y1, int yEnd,
+            double z00, double z01, double z10, double z11
+        ) {
+            int dy = y1 - y0;
+            int dx = x1 - x0;
+            // First iteration of each of the for loops below has been
+            // split into an explicit block.  This avoids a few instructions
+            // when no interpolation is necessary, but more importantly it
+            // allows us to handle dx==0 and dy==0 with no divide-by-zero
+            // problems.
+            { // y=y0
+                auto rowIter =
+                    img.x_at(x0 - img.getX0(), y0 - img.getY0());
+                { // x=x0
+                    functor(*rowIter, z00);
+                    ++rowIter;
+                }
+                for (int x = x0 + 1; x < xEnd; ++x, ++rowIter) {
+                    functor(*rowIter, ((x1 - x)*z00 + (x - x0)*z10)/dx);
+                }
+            }
+            for (int y = y0 + 1; y < yEnd; ++y) {
+                auto rowIter =
+                    img.x_at(x0 - img.getX0(), y - img.getY0());
+                double z0 = ((y1 - y)*z00 + (y - y0)*z01)/dy;
+                double z1 = ((y1 - y)*z10 + (y - y0)*z11)/dy;
+                { // x=x0
+                    functor(*rowIter, z0);
+                    ++rowIter;
+                }
+                for (int x = x0 + 1; x < xEnd; ++x, ++rowIter) {
+                    functor(*rowIter, ((x1 - x)*z0 + (x - x0)*z1)/dx);
+                }
+            }
+        };
+
+        auto processRowBlock = [&region, &field, &interpolate, xStep](int y0, int y1, int yEnd) {
+            int x0 = region.getBeginX();
+            int x1 = x0 + xStep;
+            double z00 = field.evaluate(x0, y0);
+            double z01 = field.evaluate(x0, y1);
+            double z10 = 0.0;
+            double z11 = 0.0;
+            while (x1 < region.getEndX()) {
+                z10 = field.evaluate(x1, y0);
+                z11 = field.evaluate(x1, y1);
+                interpolate(x0, x1, x1, y0, y1, yEnd, z00, z01, z10, z11);
+                x0 = x1;
+                x1 += xStep;
+                z00 = z10;
+                z01 = z11;
+            }
+            {
```

Comment:
It's conceptually a special-cased last iteration of the loop, and I like to do that at the same/scope indentation level as the rest of the loop body.  I'll add a comment to that effect.